libazureinit/src/health.rs

```diff
@@ -0,0 +1,310 @@
+    );
+    headers
+        .insert("content-type", HeaderValue::from_static("application/json"));
+
+    tracing::info!(?headers, "Prepared HTTP headers");
+
+    let connect_timeout = Duration::from_secs_f64(cfg.connection_timeout_secs);
+    let read_timeout = Duration::from_secs_f64(cfg.read_timeout_secs);
+    let retry_for = Duration::from_secs_f64(cfg.total_retry_timeout_secs);
+
+    let client = Client::builder()
+        .connect_timeout(connect_timeout)
+        .timeout(read_timeout)
+        .build()?;
+
+    let mut remaining = retry_for;
+    while !remaining.is_zero() {
+        let (resp, new_remaining) = http::post(
+            &client,
+            headers.clone(),
+            body.clone(),
+            read_timeout,
+            connect_timeout,
+            remaining,
+            &cfg.health_endpoint,
+        )
+        .await?;
+
+        let status = resp.status();
+        tracing::info!(
+            target: "libazureinit::health::status",
+            "Wireserver replied with status {}",
+            status
+        );
+
+        if status.is_success() {
+            tracing::info!(
+                target: "libazureinit::health::status",
+                "Report '{}' succeeded",
+                state
+            );
+            return Ok(());
+        }
+
+        if status == StatusCode::TOO_MANY_REQUESTS
+            || status == StatusCode::SERVICE_UNAVAILABLE
+        {
+            if let Some(retry_after) = resp.headers().get("Retry-After") {
+                if let Some(delay_secs) = retry_after
+                    .to_str()
+                    .ok()
+                    .and_then(|s| s.parse::<u64>().ok())
+                {
+                    tracing::warn!(
+                        "Retryable HTTP status {} received with Retry-After: {}s. Sleeping before retrying...",
+                        status,
+                        delay_secs
+                    );
+                    tokio::time::sleep(Duration::from_secs(delay_secs)).await;
+                } else {
+                    tracing::warn!(
+                        "Retryable HTTP status {} received. No valid Retry-After, retrying...",
+                        status
+                    );
+                }
+            } else {
+                tracing::warn!(
+                    "Retryable HTTP status {} received. No Retry-After header, retrying...",
+                    status
+                );
+            }
+        } else {
+            tracing::error!(
+                "Non-retryable HTTP status {}, bailing out",
+                status
+            );
+            return Err(Error::HttpStatus {
+                endpoint: cfg.health_endpoint.clone(),
+                status,
+            });
+        }
+
+        remaining = new_remaining;
+    }
+
+    tracing::warn!("Report '{}' timed out", state);
+    Err(Error::Timeout)
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I recommend adding a test case to cover `Retry-After` as well as a variant with a buggy/odd `Retry-After` that's greater than `total_retry_timeout_secs`